File: server/src/main/java/com/preponderous/roam/model/LivingEntity.java
package com.preponderous.roam.model;

import java.util.ArrayList;
import java.util.List;

/**
 * Represents a living entity in the game that has energy and can interact with food.
 * 
 * @author Daniel McCoy Stephenson
 */
public abstract class LivingEntity extends Entity {
    public static final double MIN_ENERGY = 0.0;
    public static final double MAX_ENERGY = 100.0;
    
    private double energy;
    private double targetEnergy;
    private final long tickCreated;
    private Long tickLastReproduced;
    private final List<Class<? extends Entity>> edibleEntityTypes;

    public LivingEntity(String name, String imagePath, double energy, long tickCreated) {
        super(name, imagePath);
        this.energy = energy;
        this.targetEnergy = energy;
        this.tickCreated = tickCreated;
        this.tickLastReproduced = null;
        this.edibleEntityTypes = new ArrayList<>();
    }

    public double getEnergy() {
        return energy;
    }

    public void setEnergy(double energy) {
        if (energy < MIN_ENERGY) {
            this.energy = MIN_ENERGY;
        } else if (energy > MAX_ENERGY) {
            this.energy = MAX_ENERGY;
        } else {
            this.energy = energy;
        }
    }

    public void addEnergy(double amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount must be non-negative: " + amount);
        }
        if (this.energy + amount > MAX_ENERGY) {
            this.energy = MAX_ENERGY;
        } else {
            this.energy += amount;
        }
    }

    public void removeEnergy(double amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount must be non-negative: " + amount);
        }
        if (this.energy - amount < MIN_ENERGY) {
            this.energy = MIN_ENERGY;
        } else {
            this.energy -= amount;
        }
    }

    public boolean needsEnergy() {
        return energy < targetEnergy * 0.50;
    }

    public double getTargetEnergy() {
        return targetEnergy;
    }

    public void setTargetEnergy(double targetEnergy) {
        this.targetEnergy = targetEnergy;
    }

    public long getTickCreated() {
        return tickCreated;
    }

    public Long getTickLastReproduced() {
        return tickLastReproduced;
    }

    public void setTickLastReproduced(Long tick) {
        this.tickLastReproduced = tick;
    }

    public long getAge(long currentTick) {
        return currentTick - tickCreated;
    }

    public void kill() {
        this.energy = 0;
    }

    public List<Class<? extends Entity>> getEdibleEntityTypes() {
        return edibleEntityTypes;
    }

    public boolean canEat(Entity entity) {
        for (Class<? extends Entity> entityType : edibleEntityTypes) {
            if (entityType.isInstance(entity)) {
                return true;
            }
        }
        return false;
    }
}

File: src/entity/living/livingEntity.py
from entity.drawableEntity import DrawableEntity


# @author Daniel McCoy Stephenson
# @since August 5th, 2022
class LivingEntity(DrawableEntity):
    # Excrement spawning cooldown: 5 minutes at 30 ticks per second
    EXCREMENT_COOLDOWN_TICKS = 30 * 60 * 5  # 9000 ticks
    
    def __init__(self, name, color, energy, edibleEntityTypes, tickCreated):
        DrawableEntity.__init__(self, name, color)
        self.energy = energy
        self.edibleEntityTypes = edibleEntityTypes
        self.targetEnergy = energy
        self.tickCreated = tickCreated
        self.tickLastReproduced = None
        self.tickLastExcrement = None

    def getEnergy(self):
        return self.energy

    def setEnergy(self, energy):
        if energy < 0:
            self.energy = 0
        elif energy > 100:
            self.energy = 100
        else:
            self.energy = energy

    def addEnergy(self, amount):
        if self.energy + amount > 100:
            self.energy = 100
        else:
            self.energy += amount

    def removeEnergy(self, amount):
        if self.energy - amount < 0:
            self.energy = 0
        else:
            self.energy -= amount

    def needsEnergy(self):
        return self.energy < self.targetEnergy * 0.50

    def getTargetEnergy(self):
        return self.targetEnergy

    def setTargetEnergy(self, targetEnergy):
        self.targetEnergy = targetEnergy

    def canEat(self, entity):
        for entityType in self.edibleEntityTypes:
            if type(entity) is entityType:
                return True
        return False

    def kill(self):
        self.energy = 0

    def getTickCreated(self):
        return self.tickCreated

    def setTickCreated(self, tick):
        self.tickCreated = tick

    def getAge(self, tick):
        return tick - self.tickCreated

    def getTickLastReproduced(self):
        return self.tickLastReproduced

    def setTickLastReproduced(self, tick):
        self.tickLastReproduced = tick

    def getTickLastExcrement(self):
        return self.tickLastExcrement

    def setTickLastExcrement(self, tick):
        self.tickLastExcrement = tick

    def shouldSpawnExcrement(self, currentTick):
        # Spawn excrement every 5 minutes (at 30 ticks per second)
        if self.tickLastExcrement is None:
            return True
        return (currentTick - self.tickLastExcrement) >= self.EXCREMENT_COOLDOWN_TICKS
